### ISSUE TITLE
Multiple KEDA deployments in a single cluster

### ISSUE THREAD

**Issue by tomkerkhove** [kedacore/keda#2654]
Over the past weeks/months I've had more and more questions if you can deploy KEDA multiple times in a single cluster.

The scenarios for this is:
1. **Hard Multi-tenancy** - People want to isolate KEDA across teams/customers in a single cluster
  - Customer A should not be able to see/interfer what customer B does
2. **Soft Multi-tenancy** - People want to isolate KEDA across teams/customers in a single cluster for management purposes ([example](https://kubernetes.slack.com/archives/CKZJ36A5D/p1642667866107800))
  - Given it's shared in the cluster, it lacks clear ownership
3. **Security** - People want to reduce the amount of access that a given KEDA deployment has
  - For example, when using pod identity a centralized KEDA needs access to every resource that the apps depend on in the cluster which is a security risk

**This issue is used to gather all scenarios to see how we can best tackle this ask**; if we start doing it. If you do have more scenarios, do let us know.

## Background - Why is this not supported today?

As per https://github.com/kedacore/keda/issues/470, you can see that Kubernetes upstream only allows end-users to run a single metric server per cluster. There is [an open proposal](https://github.com/kubernetes-sigs/custom-metrics-apiserver/issues/70) by @zroubalik to improve this, but until then we cannot deploy the metric server multiple times.

The KEDA operator allows you to watch all namespaces, or just a single one. In theory, we could change KEDA so that we can re-use the metric server and change our Helm charts to be more flexible by providing 3 of them (`metric-server`, `operator` & `full`), we still have concerns with this scenario:
- It will become harder to diagnose issues, because there is no guarantee that the operator runs the same version as the metric server
- It does not provide a full isolation between everyone. This does not fix scenario 1)
- The metric server is still a shared component, so there still has to be somebody in charge of managing it. This does not fix scenario 2)
- If somebody removes the metric server, it will impact other teams relying on it.

Because of this, we are waiting to support this model until https://github.com/kubernetes-sigs/custom-metrics-apiserver/issues/70 is tackled in KEDA upstream.

**Comment by luisg707** [comment]
+1 on this ask.

**Comment by dgrisonnet** [comment]
Ack, just wanted to make sure.

We should definitely look into having the router implemented for that purpose. Out of curiosity, do you know if the various adapters are planning to be converted to KEDA scalers in the future or some of them wants to remain standalone projects?

By the way, I started looking at the router again and I am looking for contributors to drive and maintain it. Do you perhaps have some contributors that would be interested and have time to look into that?

**Comment by zroubalik** [comment]
> Ack, just wanted to make sure.

👍 

> 
> We should definitely look into having the router implemented for that purpose. Out of curiosity, do you know if the various adapters are planning to be converted to KEDA scalers in the future or some of them wants to remain standalone projects?

The only one I am aware of is DataDog https://docs.datadoghq.com/containers/cluster_agent/external_metrics/?tab=helm, but there might be other adapters, I just don't know :)

> By the way, I started looking at the router again and I am looking for contributors to drive and maintain it. Do you perhaps have some contributors that would be interested and have time to look into that?

That's great! @kedacore/keda-contributors ^ anybody interested. I'd be interested, but the problem for me is time :( So I'd join this effort, but cannot assure full commitment.

**Comment by tomkerkhove** [comment]
I'd be happy to help on the non-code side if there is need for it.

With regards to existing adapters, Prometheus is another that is already a scaler. But I think this is not related in the sense that we should be open to support other metric servers next to KEDA as well though.

**Comment by dgrisonnet** [comment]
> With regards to existing adapters, Prometheus is another that is already a scaler. But I think this is not related in the sense that we should be open to support other metric servers next to KEDA as well though.

Yeah for Prometheus we are looking at deprecating the project in favor of KEDA Prometheus scaler which is why I was curious if other adapters also choose this path. But I completely agree that it shouldn't be a reason to limit the possibilities for the users.

**Comment by zroubalik** [comment]
It is also worth mentioning, that with the lastest architecture changes we have done (slimming down the Metrics Server logic), we might be able to provide a solution for multitenancy on our own. 
The Metrics Server will be just a light layer that will route the requests to individual KEDA installations (based on the namespace).

**Comment by JorTurFer** [comment]
I'm interested but I have limited bandwidth and I have several things I wanna do in KEDA for next weeks. After them, I'm willing to contribute/maintain this feature.
Could this match with your plans @dgrisonnet?

**Comment by vcardenas** [comment]
Hello, to mention my use case in case it may help... cluster operators like keda are managed by a platform team (or cluster admin team), but they only concern installation and upgrades, all of the ScaledObjects and more importantly, TriggerAuthentications, would be managed by dev teams (tenants) that can work within one or multiple namespaces only, but definitely not the keda namespace. That is impossible today with the need of having scalers for aws cloud resources, since the keda operator needs to either assume a role with permissions crossing tenant boundaries,  or be able to assume different roles also crossing boundaries in the process. 
The [external secrets](https://external-secrets.io/v0.7.2/guides/multi-tenancy/) operator solves this beautifully with the **Managed SecretStore per Namespace** scenario, although other multitenancy scenarios are supported. a **SecretStore** seems to be reminiscent of what a **TriggerAuthentication** is expected to be. That may be worth taking a look for ideas.

**Comment by JorTurFer** [comment]
Hi @vcardenas .
Sadly, they are different use cases. The current limitation is that metrics apis only allow a single service behind the api endpoint, which totally limits the architecture to have a single metrics server with cluster scope and this is a k8s limitation. Those projects you mentioned don't have this problem because they don't expose any kind of metric to the cluster api server.

In KEDA we are working on having this multitenant support or at least something similar but it's still in progress

**Comment by vcardenas** [comment]
I get that, my comment was oriented towards the scenario of having a single keda operator/metrics-server and multiple tenants making use of it.

**Comment by sidharthkumarpradhan** [comment]
Hello Team, Thanks to @JorTurFer for assisting me with some of my queries. Thank you so much. What I am struggling with right now is how to make KEDA watch multiple namespaces but not all. I tried to deploy multiple keda-operators by changing the WATCH_NAMESPACE values according to my need, but this is not working. Only one namespace is actively working others are not scaling up the resources. Any guidance on configuration and anything I could do to achieve my goal. Will grateful for any help here.

![image](https://github.com/kedacore/keda/assets/66263561/3b6cba26-a91f-477f-9ab0-4f76f16da058)

**Comment by JorTurFer** [comment]
as we are already talking about this on the discussion, let's continue there to not generate noise on this issue. This issue is to track the feature, and currently it's not supported.
